source/server/listener_manager_impl.cc

```diff
@@ -281,111 +292,215 @@ bool ListenerImpl::isWildcardServerName(const std::string& name) {
+    ApplicationProtocolsMap& application_protocols_map,
     const std::vector<std::string>& application_protocols,
     const Network::FilterChainSharedPtr& filter_chain) {
   if (application_protocols.empty()) {
-    transport_protocol_map[EMPTY_STRING] = filter_chain;
+    application_protocols_map[EMPTY_STRING] = filter_chain;
   } else {
     for (const auto& application_protocol : application_protocols) {
-      transport_protocol_map[application_protocol] = filter_chain;
+      application_protocols_map[application_protocol] = filter_chain;
+    }
+  }
+}
+
+void ListenerImpl::finishFilterChain() {
+  for (auto& port : filter_chains_) {
+    auto& destination_ips_pair = port.second;
+    auto& destination_ips_map = destination_ips_pair.first;
+    std::vector<std::pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>> list;
+    for (const auto& entry : destination_ips_map) {
+      std::vector<Network::Address::CidrRange> subnets;
+      if (entry.first == EMPTY_STRING) {
+        list.push_back(
+            std::make_pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>(
+                std::make_shared<ServerNamesMap>(entry.second),
+                {Network::Address::CidrRange::create("0.0.0.0/0"),
+                 Network::Address::CidrRange::create("::/0")}));
+      } else {
+        list.push_back(
+            std::make_pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>(
+                std::make_shared<ServerNamesMap>(entry.second),
+                {Network::Address::CidrRange::create(entry.first)}));
+      }
     }
+    destination_ips_pair.second.reset(
+        new Network::LcTrie::LcTrie<ServerNamesMapSharedPtr>(list, true));
   }
 }
 
 const Network::FilterChain*
 ListenerImpl::findFilterChain(const Network::ConnectionSocket& socket) const {
+  const auto& address = socket.localAddress();
+
+  // Match on destination port (only for IP addresses).
+  if (address->type() == Network::Address::Type::Ip) {
+    const auto port_match = filter_chains_.find(address->ip()->port());
+    if (port_match != filter_chains_.end()) {
+      return findFilterChainForDestinationIP(*port_match->second.second, socket);
+    }
+  }
+
+  // Match on catch-all port 0.
+  const auto port_match = filter_chains_.find(0);
+  if (port_match != filter_chains_.end()) {
+    return findFilterChainForDestinationIP(*port_match->second.second, socket);
+  }
+
+  return nullptr;
+}
+
+const Network::FilterChain*
+ListenerImpl::findFilterChainForDestinationIP(const DestinationIPsTrie& destination_ips_trie,
+                                              const Network::ConnectionSocket& socket) const {
+  auto address = socket.localAddress();
+
+  // Use invalid IP address (matching only filter chains without IP requirements) for UDS.
+  if (address->type() != Network::Address::Type::Ip) {
+    address = Network::Utility::parseInternetAddress("255.255.255.255");
```

Comment:
Can we create a static object to use for this, instead of parsing and allocating for each connection?

Reply:
Done.